scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,391 @@ def recv(self, x=0xffff):
+        if isotp_pci == 3 and isotp_fc in [0, 1, 2]:
+            if verbose:
+                print("[+] Found flow-control frame from identifier 0x%03x" %
+                      packet.identifier)
+            if isinstance(id_list, dict):
+                id_list[id_value] = (packet, packet.identifier)
+            if isinstance(id_list, list):
+                id_list.append(id_value)
+        else:
+            noise_ids.append(packet.identifier)
+    except Exception as e:
+        print("[!] Unknown message Exception: " + str(e) + " on packet:" +
+              packet.__repr__())
+
+
+def scan(sock, scan_range=range(0x800), noise_ids=None, sniff_time=0.1,
+         verbose=False):
+    """Scan and return dictionary of detections
+
+    Args:
+            sock: socket for can interface
+            scan_range: hexadecimal range of IDs to scan.
+                        Default is 0x0 - 0x7ff
+            noise_ids: list of packet IDs which will not be considered when
+                       received during scan
+            sniff_time: time the scan waits for isotp flow control responses
+                        after sending a first frame
+            verbose: displays information during scan
+
+    ISOTP-Scan - NO extended IDs
+    found_packets = Dictionary with Send-to-ID as
+    key and a tuple (received packet, Recv_ID)
+    """
+    return_values = dict()
+    for value in scan_range:
+        sock.sniff(prn=lambda pkt: get_isotp_fc(value, return_values,
+                                                noise_ids, False, pkt,
+                                                verbose),
+                   timeout=sniff_time,
+                   started_callback=lambda: sock.send(
+                       get_isotp_packet(value)))
+    return return_values
+
+
+def scan_extended(sock, scan_range=range(0x800), scan_block_size=100,
+                  noise_ids=None, sniff_time=0.1, verbose=False):
+    """Scan with extended addresses and return dictionary of detections
+
+    Args:
+            sock: socket for can interface
+            scan_range: hexadecimal range of IDs to scan.
+                        Default is 0x0 - 0x7ff
+            scan_block_size: count of packets send at once
+            noise_ids: list of packet IDs which will not be considered when
+                       received during scan
+            sniff_time: time the scan waits for isotp flow control responses
+                        after sending a first frame
+            verbose: displays information during scan
+
+    If an answer-packet found -> slow scan with
+    single packages with extended ID 0 - 255
+    found_packets = Dictionary with Send-to-ID
+    as key and a tuple (received packet, Recv_ID)
+    """
+
+    return_values = dict()
+    for value in scan_range:
+        pkt = get_isotp_packet(value, extended=True)
+        id_list = []
+        scan_block_size = 1 if scan_block_size == 0 else scan_block_size
```

Comment:
I don't think the `scan_block_size` has to be set in every iteration. Once at the beginning should be enough. So maybe move it before the `for` line

Reply:
done